documentation/source/protocol_extension.rst

```diff
@@ -0,0 +1,90 @@
+How to write a protocol extension
+=================================
+
+Cocotb take care of the core communication between the simulator and testbench.
+To use some known busses like wishbone, avalon, spi, i2c,... An extension is required.
+
+This is a guideline that will explain how to write a python module protocol
+extension for cocotb.
+
+Naming convention
+-----------------
+
+A cocotb extension module must begin by `cocotb_` and terminate with protocol
+name.
+
+For exemple, with spi the module extension is `cocotb_spi`.
+
+Bus driver
+----------
+
+A Bus driver is a python class that factorize common behavior and check for
+presence of some signals.
+
+A Bus driver is an active class that be driven in the main testbench. Example is
+given with avalon bus in `cocotb/drivers/avalon.py`.
+
+With AvalonMaster class we can generate stimulis for reading and writing on bus.
+
+The main class is named `BusDriver` and is defined in `cocotb/drivers/__init__.py`
+
+TODO: give an example.
+
+Bus Monitor
+-----------
+
+A Bus monitor is mainly a passive class that will monitor the bus.
+
+TODO: explain monitor
+
+Python module
+-------------
+
+The module file hierarchy should be as following ::
+
+  ├── cocotb_spi/
+  │   └── __init__.py
+  ├── README.md
+  └── setup.py
+
+With python module source code in directory cocotb_spi and module configuration
+in setup.py
+
+setup.py is a python configuration file that be used for packaging. An example
+is given above ::
+
+  try:
+    from setuptools import setup
+  except ImportError:
+    from distutils.core import setup
+
+  setup(name='cocotb_spi',
+      version='0.1',
+      py_modules=['cocotb_spi'],
+      )
+
+With this config file, python module can be installed with ::
+
+  $ python setup.py install
+
+or for development::
+
+  $ python setup.py develop
+
+
+Then module can be imported with import keyword in your testbench ::
+
+  import cocotb_spi
```

Comment:
An alternative would be `cocotb.ext.spi`

Reply:
If it's possible with an external package it could be better yes

Reply:
PEP420 makes it easy to make `cocotb.ext` a namespace package, which allows multiple distributions (things with a `setup.py`) to exist within the same package - all that is required is that third-party extensions do _not_ include a `cocotb/__init__.py` or a `cocotb/ext/__init__.py`.

Reply:
have you got a template example for that @eric-wieser ? I'm trying to do that without success [here]( https://github.com/Martoni/cocotb_ext_spi.git).

Reply:
It's looking like that might be a bad idea after all. I'll try to collect some example of other projects choosing to do this here:

* `flask.ext` was removed: https://github.com/pallets/flask/issues/1135#issuecomment-61860862

Reply:
Ok, glad to hear that. It was really painful to do (without success !)

Reply:
A much simpler option would be `cocotb_ext.spi`, but I'm not sure that's an improvement over `cocotb_spi`